accepted/0034-context-config-parameter.md

```diff
@@ -0,0 +1,116 @@
+
+There do not appear to be many references to variables named `context` so these should be reasonably straight forward.
+
+### `npm config`
+
+Currently we create an implicit `where` variable in the implementation seen [here](https://github.com/npm/cli/blob/8806015fdd025f73ccf4001472370eafd3c5a856/lib/config.js#L122) with one of two values, dependent on the `global` flag. The implementation would involve creating `context` as a variable in the config, then modifying the flattener for `global` such that when set the `context` is changed to `'global'`, then using the `context` as the source of truth when determining which file to write.
+
+By default `context` will have no value, this is how we will avoid a breaking change.
+
+The `adduser` and `logout` commands currently are hardcoded to use `'user'` instead of referencing a variable, which is behavior I believe should remain as-is in order to avoid users accidentally saving credentials into project-level config files.
+
+### `npm exec`
+
+Implementation here is somewhat more involved. The logic found [here](https://github.com/npm/cli/blob/8806015fdd025f73ccf4001472370eafd3c5a856/lib/exec.js#L99-L117) would be expanded to create an instance of arborist, load your actual tree and iterate its inventory looking for a bin based on the rules detailed above.
+
+For project installed packages, if a bin is found, we would verify that the link and/or shim located in that location's bin directory points to the exact package we matched. If it does not, we would overwrite it with one that does, and then execute the freshly linked/shimmed bin.
+
+For the case of globally installed packages, there is no need to ensure the correct bin is linked as we will only link the bin of a directly installed global package. However, we would still apply the search logic here so that `npm exec --context=global -- typescript` will function as expected.
+
+Remote package behavior would remain exactly as it is today.
+
+When the `context` config is unset (the default), we will maintain the existing behavior of searching locally installed packages, then globally installed packages, then remote packages. The only change would be that the bin we run will be overwritten to the shallowest match in the user's actual tree before executing it, eliminating the current race condition that exists in the case of a conflicting bin in your tree.
+
+There is the potential that when run in interactive mode we can also display a list of everywhere we found something we _could_ run based on the user's request and prompt them to tell us which one.
+
+## Unresolved Questions and Bikeshedding
+
+~Is `where` the best name for this config? What if there are too many conflicts in dependencies?~ EDIT: Document updated to use `context` instead of `where`.
+
+What would the shortcut be? `-c` is currently the short form of `--call`.
+
+### `npm config`
+
+Should we prevent a user from setting protected fields like `_auth` or `_token` to a local config?
```

Comment:
Let's stay consistent with whatever is supported by `npm config` by default &/or when run in the `global` _context_. If we support different behaviour in the `user` or `global` _contexts_ today I'd wonder why; The whole point of unlocking `local`/project-level _context_ is to eliminate the need to manually manage an `.npmrc` file. Limiting the fields you can `set` would also reduce the usability of defining the `local` _context_

Reply:
that's a fair point :+1: